mv: gossip the same backlog if a different backlog was sent in a response

wmitros opened the pull request:
Currently, there are 2 ways of sharing a backlog with other nodes: through
a gossip mechanism, and with responses to replica writes. In gossip, we
check each second if the backlog changed, and if it did we update other
nodes with it. However if the backlog for this node changed on another
node with a write response, the gossiped backlog is currently not updated,
so if after the response the backlog goes back to the value from the previous
gossip round, it will not get sent and the other node will stay with an
outdated backlog - this can be observed in the following scenario:

1. Cluster starts, all nodes gossip their empty view update backlog to one another
2. On node N, `view_update_backlog_broker` (the backlog gossiper) performs an iteration of its backlog update loop, sees no change (backlog has been empty since the start), schedules the next iteration after 1s
3. Within the next 1s, coordinator (different than N) sends a write to N causing a remote view update (which we do not wait for). As a result, node N replies immediately with an increased view update backlog, which is then noted by the coordinator.
4. Still within the 1s, node N finishes the view update in the background, dropping its view update backlog to 0.
5. In the next and following iterations of `view_update_backlog_broker` on N, backlog is empty, as it was in step 2, so no change is seen and no update is sent due to the check
```
auto backlog = _sp.local().get_view_update_backlog(); 
if (backlog_published && *backlog_published == backlog) { 
    sleep_abortable(gms::gossiper::INTERVAL, _as).get(); 
    continue; 
} 
```

After this scenario happens, the coordinator keeps an information about an increased view update backlog on N even though it's actually already empty

This patch fixes the issue this by notifying the gossip that a different backlog
was sent in a response, causing it to send an unchanged backlog to other
nodes in the following gossip round.

Fixes: https://github.com/scylladb/scylladb/issues/18461

Similarly to https://github.com/scylladb/scylladb/pull/18646, without admission control (https://github.com/scylladb/scylladb/pull/18334), this patch doesn't affect much, so I'm marking it as backport/none

Tests: manual. Currently this patch only affects the length of MV flow control delay, which is not reliable to base a test on. A proper test will be added when MV admission control is added, so we'll be able to base the test on rejected requests

wmitros commented:
This patch only concerns the last commit. Is based on https://github.com/scylladb/scylladb/pull/18646 to avoid later conflicts

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8780/)
#### Build Details:
- Duration: 40 min
- Builder: i-032a01f421ac12cb1 (r5ad.8xlarge)

wmitros commented:
Rebased on updated https://github.com/scylladb/scylladb/pull/18646 and fixed compilation error

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8818/)
#### Build Failure:
```
[2024-05-15T10:43:13.261Z] FAILED: build/debug/replica/database.o 
[2024-05-15T10:43:13.261Z] clang++ -MD -MT build/debug/replica/database.o -MF build/debug/replica/database.o.d -std=c++20 -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/gen/include -U_FORTIFY_SOURCE -Werror=unused-result -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_SSTRING -DSEASTAR_LOGGER_COMPILE_TIME_FMT -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_DEBUG -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_DEBUG_PROMISE -DSEASTAR_LOGGER_TYPE_STDOUT -DSEASTAR_TYPE_ERASE_MORE -DBOOST_NO_CXX98_FUNCTION_BASE -DFMT_SHARED -I/usr/include/p11-kit-1 -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DDEBUG -DSANITIZE -DDEBUG_LSA_SANITIZER -DSCYLLA_ENABLE_ERROR_INJECTION -Og -DSCYLLA_BUILD_MODE=debug -g -gz -iquote. -iquote build/debug/gen -std=gnu++20  -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DBOOST_ALL_DYN_LINK    -fvisibility=hidden -isystem abseil -Wall -Werror -Wextra -Wimplicit-fallthrough -Wno-mismatched-tags -Wno-c++11-narrowing -Wno-overloaded-virtual -Wno-unused-parameter -Wno-unsupported-friend -Wno-missing-field-initializers -Wno-deprecated-copy -Wno-psabi -Wno-enum-constexpr-conversion -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN  -c -o build/debug/replica/database.o replica/database.cc
[2024-05-15T10:43:13.263Z] replica/database.cc:2026:55: error: get_view_update_backlog_for_response is a private member of service::storage_proxy
[2024-05-15T10:43:13.264Z]         backlog = _view_update_generator->get_proxy().get_view_update_backlog_for_response();
[2024-05-15T10:43:13.264Z]                                                       ^
[2024-05-15T10:43:13.264Z] ./service/storage_proxy.hh:442:30: note: declared private here
[2024-05-15T10:43:13.264Z]     db::view::update_backlog get_view_update_backlog_for_response() const;
[2024-05-15T10:43:13.264Z]                              ^
[2024-05-15T10:43:13.264Z] 1 error generated.
```
#### Build Details:
- Duration: 34 min
- Builder: i-01f580a432a2fd8cf (m5ad.8xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8851/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8851/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/batchlog_manager_test
:small_blue_diamond: boost/database_test
:small_blue_diamond: boost/repair_test
:small_blue_diamond: topology_custom/test_mv_backlog
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8851/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9618/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1597/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8851/)
#### Build Details:
- Duration: 5 hr 4 min
- Builder: spider8.cloudius-systems.com

piodul commented on service/storage_proxy.cc:
```diff
@@ -2447,8 +2447,12 @@ void storage_proxy::maybe_update_view_backlog_of(gms::inet_address replica, std:
+    return _max_view_update_backlog.add_fetch_for_response(this_shard_id(), get_db().local().get_view_update_backlog());
+}
+
+std::optional<db::view::update_backlog> storage_proxy::get_view_update_backlog_for_gossip(std::optional<db::view::update_backlog> last_backlog) const {
+    return _max_view_update_backlog.add_fetch_for_gossip(this_shard_id(), get_db().local().get_view_update_backlog(), std::move(last_backlog));
 }
```
Gossiper runs on shard 0, so the shard parameter is unneeded. It would make more sense to check the shard and do on_internal_error is it is not 0.

wmitros replied:
Added

piodul commented on service/storage_proxy.hh:
```diff
@@ -480,7 +480,9 @@ public:
-    db::view::update_backlog get_view_update_backlog() const;
+    db::view::update_backlog get_view_update_backlog_for_response() const;
+
+    std::optional<db::view::update_backlog> get_view_update_backlog_for_gossip(std::optional<db::view::update_backlog> last_backlog) const;
```
I don't like these names because they don't really tell much about what those functions do. It's not obvious what are the requirements of "response" or "gossip".

wmitros replied:
I changed the names a bit

piodul commented on db/view/view.cc:
```diff
@@ -2626,6 +2626,27 @@ update_backlog node_update_backlog::add_fetch(unsigned shard, update_backlog bac
+        need_publishing = np.exchange(need_publishing::no, std::memory_order_relaxed);
+    }
+    auto max = add_fetch(shard, backlog);
+    if (!need_publishing && last_backlog && *last_backlog == max) {
+        return std::nullopt;
+    } else {
+        return max;
+    }
+}
```
Do we have to check agains the last backlog at all? If the value has changed then one of the `need_publishing` flags must have been updated, no?

wmitros commented on db/view/node_view_update_backlog.hh:
```diff
@@ -37,16 +39,20 @@ class node_update_backlog {
     std::chrono::milliseconds _interval;
     std::atomic<clock::time_point> _last_update;
     std::atomic<update_backlog> _max;
+    std::vector<std::atomic<need_publishing>> _need_publishing;
```
Added

wmitros commented on db/view/view.cc:
```diff
@@ -2626,6 +2626,27 @@ update_backlog node_update_backlog::add_fetch(unsigned shard, update_backlog bac
+        }
+        need_publishing = np.exchange(need_publishing::no, std::memory_order_relaxed);
+    }
+    auto max = add_fetch(shard, backlog);
```
Makes sense

piodul commented:
> However if the backlog for this node changed on another
> node with a write response, the gossiped backlog is currently not updated,
> so if after the response the backlog goes back to the value from the previous
> gossip round, it will not get sent and the other node will stay with an
> outdated backlog.

Your description is very information-dense and thus difficult to parse. In such cases, an example usually helps in understanding the bug. The example that you already provided in the description of the issue was, in my opinion, quite clear, so you could write an example here based on that.

wmitros commented:
> Your description is very information-dense and thus difficult to parse. In such cases, an example usually helps in understanding the bug. The example that you already provided in the description of the issue was, in my opinion, quite clear, so you could write an example here based on that.

I've mostly copied the example, hopefully that's good enough.

Also I rebased to fix the merge conflicts. At the same time, I added a commit for better readability to this PR so the patch is now the last 2 commits, not only the last one

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9051/)
#### Build Details:
- Duration: 4 hr 36 min
- Builder: i-0656e49f3aa4a8738 (m5d.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9061/)
#### Build Failure:
```
[2024-05-22T08:02:37.215Z] FAILED: build/dev/db/view/node_view_update_backlog.hh.o 
[2024-05-22T08:02:37.215Z] clang++ -MD -MT build/dev/db/view/node_view_update_backlog.hh.o -MF build/dev/db/view/node_view_update_backlog.hh.o.d -std=c++20 -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/gen/include -U_FORTIFY_SOURCE -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_SSTRING -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_LOGGER_COMPILE_TIME_FMT -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_LOGGER_TYPE_STDOUT -DSEASTAR_TYPE_ERASE_MORE -DBOOST_NO_CXX98_FUNCTION_BASE -DFMT_SHARED -I/usr/include/p11-kit-1 -std=gnu++20  -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DBOOST_ALL_DYN_LINK    -fvisibility=hidden -isystem abseil -Wall -Werror -Wextra -Wimplicit-fallthrough -Wno-mismatched-tags -Wno-c++11-narrowing -Wno-overloaded-virtual -Wno-unused-parameter -Wno-unsupported-friend -Wno-missing-field-initializers -Wno-deprecated-copy -Wno-psabi -Wno-enum-constexpr-conversion -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -DSCYLLA_ENABLE_PREEMPTION_SOURCE -O2 -DSCYLLA_BUILD_MODE=dev -iquote. -iquote build/dev/gen  --include db/view/node_view_update_backlog.hh -c -o build/dev/db/view/node_view_update_backlog.hh.o build/dev/gen/empty.cc
[2024-05-22T08:02:37.216Z] In file included from <built-in>:1:
[2024-05-22T08:02:37.216Z] ./db/view/node_view_update_backlog.hh:55:5: error: no template named future; did you mean seastar::future?
[2024-05-22T08:02:37.217Z]     future<std::optional<update_backlog>> fetch_if_changed(std::optional<db::view::update_backlog> last_backlog);
[2024-05-22T08:02:37.217Z]     ^~~~~~
[2024-05-22T08:02:37.217Z]     seastar::future
[2024-05-22T08:02:37.217Z] /jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include/seastar/core/future.hh:1242:21: note: seastar::future declared here
[2024-05-22T08:02:37.217Z] class [[nodiscard]] future : private internal::future_base {
[2024-05-22T08:02:37.217Z]                     ^
[2024-05-22T08:02:37.217Z] 1 error generated.
```
#### Build Details:
- Duration: 9 hr 46 min
- Builder: i-011fe52b636329a1d (m5d.12xlarge)

piodul commented on db/view/view.cc:
```diff
@@ -2645,6 +2646,30 @@ update_backlog node_update_backlog::fetch() {
+            });
+
+    auto max = fetch();
+    if (!np && last_backlog && *last_backlog == max) {
```
I asked about this before, but the comment was resolved.

Isn't it sufficient to just check:

```c++
if (!np) {
```

?

Previously, we used to compare the last backlog vs the current one in order to decide whether it needs publishing. Now, we have the `need_publishing` flag. Is the `last_backlog && *last_backlog == max` still necessary?

piodul replied:
The whole if..else could be written as a ternary operator:

```c++
return np ? max : std::nullopt;
```

I didn't try to compile it, there is a chance that you'll get a type error because both branches don't have the same type, however it would be nice to try it.

wmitros replied:
> I asked about this before, but the comment was resolved.
> 
> Isn't it sufficient to just check:
> 
> ```c++
> if (!np) {
> ```
> 
> ?
> 
> Previously, we used to compare the last backlog vs the current one in order to decide whether it needs publishing. Now, we have the `need_publishing` flag. Is the `last_backlog && *last_backlog == max` still necessary?

After https://github.com/scylladb/scylladb/pull/18804 is merged and the `need_publishing` is also updated in that context, the `last_backlog && *last_backlog == max` check won't be needed anymore. Without #18804 the `need_publishing` flag is not set when a backlog drops, so we have no other way of detecting it other than comparing with  the last sent backlog

wmitros replied:
I used this in the last rebase

piodul commented on db/view/view.cc:
```diff
@@ -2645,6 +2646,30 @@ update_backlog node_update_backlog::fetch() {
+    auto np = co_await map_reduce(boost::irange(0u, smp::count),
+            [this] (shard_id shard) {
+                return smp::submit_to(shard, [this, shard] {
+                    if (_backlogs[shard].need_publishing) {
+                        _backlogs[shard].need_publishing = need_publishing::no;
+                        return need_publishing::yes;
+                    }
+                    return need_publishing::no;
```
Nit: you can use `std::exchange` to make the body of the lambda nice and short:

```c++
return std::exchange(_backlogs[shard].need_publishing, need_publishing::no);
```

wmitros replied:
Ok

piodul commented on service/storage_proxy.hh:
```diff
@@ -235,6 +235,8 @@ public:
     // Get information about this node's view update backlog. It combines information from all local shards.
     db::view::update_backlog get_view_update_backlog();
 
+    future<std::optional<db::view::update_backlog>> get_view_update_backlog_if_changed(std::optional<db::view::update_backlog> last_backlog);
```
Nit: you require that this function must be called on shard 0 only. Please put this information in a comment near the signature so that the future users are not surprised.

wmitros replied:
Added

piodul commented on db/view/view.cc:
```diff
@@ -2645,6 +2646,30 @@ update_backlog node_update_backlog::fetch() {
+                return a || b;
+            });
+
+    auto max = fetch();
```
This will be prone to races. Consider the following scenario:

- `_backlog` is updated on shard 1, the `need_publishing` flag is set,
- The `_last_update` is recalculated, next one will happen 10ms from now or later,
- 5ms pass,
- `fetch_if_changed` and clears the `need_publishing` flag on shard 1, but uses `fetch` to read the old value,
- There are no MV writes anymore and we didn't publish the updated backlog on shard 1.

You should collect the backlogs in the `submit_to` lambda and calculate the maximum in the reducer lambda (you should still collect the np flag in addition to this).

wmitros replied:
Added in the rebase. This method now updates the max backlog unconditionally (disregarding the `10ms` interval - it's only run once every second anyway), and returns the new max only if the `need_publishing` flag was set somewhere

piodul reviewed (changes requested):
The description looks better, thanks.

I reviewed the last commit. I left some comments. Please fix the nits, the compilation error and, if possible, rebase on master so that this commit does not require https://github.com/scylladb/scylladb/pull/18646.

Otherwise, looks good.

wmitros commented:
I rebased on master and fixed the last few comments in the last rebases

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9495/)
#### Build Failure:
```
[2024-06-05T23:26:00.417Z] FAILED: build/dev/db/view/node_view_update_backlog.hh.o 
[2024-06-05T23:26:00.417Z] clang++ -MD -MT build/dev/db/view/node_view_update_backlog.hh.o -MF build/dev/db/view/node_view_update_backlog.hh.o.d -std=c++20 -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/gen/include -U_FORTIFY_SOURCE -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_SSTRING -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_LOGGER_COMPILE_TIME_FMT -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_LOGGER_TYPE_STDOUT -DSEASTAR_TYPE_ERASE_MORE -DBOOST_NO_CXX98_FUNCTION_BASE -DFMT_SHARED -I/usr/include/p11-kit-1 -std=gnu++20  -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DBOOST_ALL_DYN_LINK    -fvisibility=hidden -isystem abseil -Wall -Werror -Wextra -Wimplicit-fallthrough -Wno-mismatched-tags -Wno-c++11-narrowing -Wno-overloaded-virtual -Wno-unused-parameter -Wno-unsupported-friend -Wno-missing-field-initializers -Wno-deprecated-copy -Wno-psabi -Wno-enum-constexpr-conversion -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -DSCYLLA_ENABLE_PREEMPTION_SOURCE -O2 -DSCYLLA_BUILD_MODE=dev -iquote. -iquote build/dev/gen  --include db/view/node_view_update_backlog.hh -c -o build/dev/db/view/node_view_update_backlog.hh.o build/dev/gen/empty.cc
[2024-06-05T23:26:00.418Z] In file included from <built-in>:1:
[2024-06-05T23:26:00.418Z] ./db/view/node_view_update_backlog.hh:55:5: error: no template named future; did you mean seastar::future?
[2024-06-05T23:26:00.418Z]     future<std::optional<update_backlog>> fetch_if_changed();
[2024-06-05T23:26:00.418Z]     ^~~~~~
[2024-06-05T23:26:00.418Z]     seastar::future
[2024-06-05T23:26:00.418Z] /jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include/seastar/core/future.hh:1240:21: note: seastar::future declared here
[2024-06-05T23:26:00.418Z] class [[nodiscard]] future : private internal::future_base {
[2024-06-05T23:26:00.418Z]                     ^
[2024-06-05T23:26:00.418Z] 1 error generated.
```
#### Build Details:
- Duration: 30 min
- Builder: spider2.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9498/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9498/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/view_schema_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9498/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1912/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9937/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9498/)
#### Build Details:
- Duration: 3 hr 6 min
- Builder: spider2.cloudius-systems.com

piodul commented on test/boost/view_schema_test.cc:
```diff
@@ -2573,13 +2573,22 @@ SEASTAR_TEST_CASE(test_unselected_column) {
 
 SEASTAR_THREAD_TEST_CASE(node_view_update_backlog) {
-    db::view::node_update_backlog b(2, 10ms);
+    db::view::node_update_backlog b(2, 100ms);
```
Nit: why was it necessary to increase the interval?

wmitros replied:
The test seems to be executing slightly slower due to the `invoke_on`s and it was failing with the original interval. It was very rare, less than 1%, but I assume we want the fail to be even less likely than that.

piodul reviewed (approved):
Just one question, but not a critical one - LGTM, I'll go ahead and merge.

piodul commented:
Queued
